File: example/step-definitions/api.steps.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { ConductorWorld } from 'conductor';

interface TodoResponse {
  id: number;
  title: string;
  status: string;
  priority: string;
}

async function ensureApi(world: ConductorWorld): Promise<void> {
  if (!world.api.isInitialized) {
    if (world.web.isLaunched) {
      await world.api.init(world.web.context);
    } else {
      await world.api.init();
    }
  }
}

Given('a todo {string} exists via the API', async function (this: ConductorWorld, title: string) {
  await ensureApi(this);
  const response = await this.api.post(`${this.config.api.baseUrl}/todos`, { title, status: 'open' });
  if (!response.ok()) throw new Error(`Failed to create todo: ${response.status()} ${response.statusText()}`);
  const todo = await response.json() as TodoResponse;
  this.data.lastTodoId = todo.id;
  this.data.lastTodoTitle = title;
});

Given('a todo {string} with priority {string} exists via the API', async function (this: ConductorWorld, title: string, priority: string) {
  await ensureApi(this);
  const response = await this.api.post(`${this.config.api.baseUrl}/todos`, { title, status: 'open', priority });
  if (!response.ok()) throw new Error(`Failed to create todo: ${response.status()} ${response.statusText()}`);
  const todo = await response.json() as TodoResponse;
  this.data.lastTodoId = todo.id;
  this.data.lastTodoTitle = title;
});

When('I create a todo {string} via the API', async function (this: ConductorWorld, title: string) {
  await ensureApi(this);
  const response = await this.api.post(`${this.config.api.baseUrl}/todos`, { title, status: 'open' });
  if (!response.ok()) throw new Error(`Failed to create todo: ${response.status()} ${response.statusText()}`);
  const todo = await response.json() as TodoResponse;
  this.data.lastTodoId = todo.id;
  this.data.lastTodoTitle = title;
});

When('I create a todo {string} with priority {string} via the API', async function (this: ConductorWorld, title: string, priority: string) {
  await ensureApi(this);
  const response = await this.api.post(`${this.config.api.baseUrl}/todos`, { title, status: 'open', priority });
  if (!response.ok()) throw new Error(`Failed to create todo: ${response.status()} ${response.statusText()}`);
  const todo = await response.json() as TodoResponse;
  this.data.lastTodoId = todo.id;
  this.data.lastTodoTitle = title;
});

When('I update the todo {string} title to {string} via the API', async function (this: ConductorWorld, currentTitle: string, newTitle: string) {
  await ensureApi(this);
  const todos = await (await this.api.get(`${this.config.api.baseUrl}/todos`)).json() as TodoResponse[];
  const todo = todos.find(t => t.title === currentTitle);
  if (!todo) throw new Error(`Todo "${currentTitle}" not found`);
  const response = await this.api.put(`${this.config.api.baseUrl}/todos/${todo.id}`, { title: newTitle });
  if (!response.ok()) throw new Error(`Failed to update todo: ${response.status()} ${response.statusText()}`);
  this.data.lastTodoId = todo.id;
  this.data.lastTodoTitle = newTitle;
});

When('I update the todo {string} status to {string} via the API', async function (this: ConductorWorld, title: string, status: string) {
  await ensureApi(this);
  const todos = await (await this.api.get(`${this.config.api.baseUrl}/todos`)).json() as TodoResponse[];
  const todo = todos.find(t => t.title === title);
  if (!todo) throw new Error(`Todo "${title}" not found`);
  const response = await this.api.put(`${this.config.api.baseUrl}/todos/${todo.id}`, { status });
  if (!response.ok()) throw new Error(`Failed to update todo: ${response.status()} ${response.statusText()}`);
});

When('I update the todo {string} priority to {string} via the API', async function (this: ConductorWorld, title: string, priority: string) {
  await ensureApi(this);
  const todos = await (await this.api.get(`${this.config.api.baseUrl}/todos`)).json() as TodoResponse[];
  const todo = todos.find(t => t.title === title);
  if (!todo) throw new Error(`Todo "${title}" not found`);
  const response = await this.api.put(`${this.config.api.baseUrl}/todos/${todo.id}`, { priority });
  if (!response.ok()) throw new Error(`Failed to update todo: ${response.status()} ${response.statusText()}`);
});

When('I delete the todo {string} via the API', async function (this: ConductorWorld, title: string) {
  await ensureApi(this);
  const todos = await (await this.api.get(`${this.config.api.baseUrl}/todos`)).json() as TodoResponse[];
  const todo = todos.find(t => t.title === title);
  if (!todo) throw new Error(`Todo "${title}" not found`);
  const response = await this.api.delete(`${this.config.api.baseUrl}/todos/${todo.id}`);
  if (response.status() !== 204) throw new Error(`Failed to delete todo: ${response.status()} ${response.statusText()}`);
});

Then('the API should return the todo {string} with status {string}', async function (this: ConductorWorld, title: string, status: string) {
  await ensureApi(this);
  const response = await this.api.get(`${this.config.api.baseUrl}/todos`);
  if (!response.ok()) throw new Error(`API returned ${response.status()}: ${response.statusText()}`);
  const todos = await response.json() as TodoResponse[];
  const found = todos.find((t) => t.title === title);
  if (!found) throw new Error(`Todo "${title}" not found in API response`);
  if (found.status !== status) throw new Error(`Expected status "${status}" but got "${found.status}"`);
});

Then('the API should return the todo {string} with priority {string}', async function (this: ConductorWorld, title: string, priority: string) {
  await ensureApi(this);
  const response = await this.api.get(`${this.config.api.baseUrl}/todos`);
  if (!response.ok()) throw new Error(`API returned ${response.status()}: ${response.statusText()}`);
  const todos = await response.json() as TodoResponse[];
  const found = todos.find((t) => t.title === title);
  if (!found) throw new Error(`Todo "${title}" not found in API response`);
  if (found.priority !== priority) throw new Error(`Expected priority "${priority}" but got "${found.priority}"`);
});

Then('the API should not return a todo {string}', async function (this: ConductorWorld, title: string) {
  await ensureApi(this);
  const response = await this.api.get(`${this.config.api.baseUrl}/todos`);
  if (!response.ok()) throw new Error(`API returned ${response.status()}: ${response.statusText()}`);
  const todos = await response.json() as TodoResponse[];
  const found = todos.find((t) => t.title === title);
  if (found) throw new Error(`Todo "${title}" should not exist but was found`);
});

Then('the API should return {int} todo(s)', async function (this: ConductorWorld, count: number) {
  await ensureApi(this);
  const response = await this.api.get(`${this.config.api.baseUrl}/todos`);
  if (!response.ok()) throw new Error(`API returned ${response.status()}: ${response.statusText()}`);
  const todos = await response.json() as TodoResponse[];
  if (todos.length !== count) throw new Error(`Expected ${count} todos but got ${todos.length}`);
});

File: java/conductor-example/src/test/java/com/nouhouari/conductor/example/stepdefs/ApiSteps.java
package com.nouhouari.conductor.example.stepdefs;

import com.microsoft.playwright.APIResponse;
import com.nouhouari.conductor.example.model.TodoResponse;
import com.nouhouari.conductor.world.ConductorWorld;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import java.util.List;
import java.util.Map;

/** Java port of example/step-definitions/api.steps.ts. */
public class ApiSteps {

    private final ConductorWorld world;

    public ApiSteps(ConductorWorld world) {
        this.world = world;
    }

    private String todosUrl() {
        return world.config.api().baseUrl() + "/api/todos";
    }

    private void requireOk(APIResponse response, String action) {
        if (!response.ok()) {
            throw new RuntimeException("Failed to " + action + ": " + response.status() + " " + response.statusText());
        }
    }

    private TodoResponse findByTitle(String title) {
        List<TodoResponse> todos = world.api().jsonList(world.api().get(todosUrl()), TodoResponse.class);
        return todos.stream().filter(t -> t.title().equals(title)).findFirst()
                .orElseThrow(() -> new RuntimeException("Todo \"" + title + "\" not found"));
    }

    @Given("a todo {string} exists via the API")
    public void aTodoExistsViaTheApi(String title) {
        ApiHelpers.ensureApi(world);
        APIResponse response = world.api().post(todosUrl(), Map.of("title", title, "status", "open"));
        requireOk(response, "create todo");
        TodoResponse todo = world.api().json(response, TodoResponse.class);
        world.data.put("lastTodoId", todo.id());
        world.data.put("lastTodoTitle", title);
    }

    @Given("a todo {string} with priority {string} exists via the API")
    public void aTodoWithPriorityExistsViaTheApi(String title, String priority) {
        ApiHelpers.ensureApi(world);
        APIResponse response = world.api().post(todosUrl(), Map.of("title", title, "status", "open", "priority", priority));
        requireOk(response, "create todo");
        TodoResponse todo = world.api().json(response, TodoResponse.class);
        world.data.put("lastTodoId", todo.id());
        world.data.put("lastTodoTitle", title);
    }

    @When("I create a todo {string} via the API")
    public void iCreateATodoViaTheApi(String title) {
        ApiHelpers.ensureApi(world);
        APIResponse response = world.api().post(todosUrl(), Map.of("title", title, "status", "open"));
        requireOk(response, "create todo");
        TodoResponse todo = world.api().json(response, TodoResponse.class);
        world.data.put("lastTodoId", todo.id());
        world.data.put("lastTodoTitle", title);
    }

    @When("I create a todo {string} with priority {string} via the API")
    public void iCreateATodoWithPriorityViaTheApi(String title, String priority) {
        ApiHelpers.ensureApi(world);
        APIResponse response = world.api().post(todosUrl(), Map.of("title", title, "status", "open", "priority", priority));
        requireOk(response, "create todo");
        TodoResponse todo = world.api().json(response, TodoResponse.class);
        world.data.put("lastTodoId", todo.id());
        world.data.put("lastTodoTitle", title);
    }

    @When("I update the todo {string} title to {string} via the API")
    public void iUpdateTheTodoTitleToViaTheApi(String currentTitle, String newTitle) {
        ApiHelpers.ensureApi(world);
        TodoResponse todo = findByTitle(currentTitle);
        APIResponse response = world.api().put(todosUrl() + "/" + todo.id(), Map.of("title", newTitle));
        requireOk(response, "update todo");
        world.data.put("lastTodoId", todo.id());
        world.data.put("lastTodoTitle", newTitle);
    }

    @When("I update the todo {string} status to {string} via the API")
    public void iUpdateTheTodoStatusToViaTheApi(String title, String status) {
        ApiHelpers.ensureApi(world);
        TodoResponse todo = findByTitle(title);
        APIResponse response = world.api().put(todosUrl() + "/" + todo.id(), Map.of("status", status));
        requireOk(response, "update todo");
    }

    @When("I update the todo {string} priority to {string} via the API")
    public void iUpdateTheTodoPriorityToViaTheApi(String title, String priority) {
        ApiHelpers.ensureApi(world);
        TodoResponse todo = findByTitle(title);
        APIResponse response = world.api().put(todosUrl() + "/" + todo.id(), Map.of("priority", priority));
        requireOk(response, "update todo");
    }

    @When("I delete the todo {string} via the API")
    public void iDeleteTheTodoViaTheApi(String title) {
        ApiHelpers.ensureApi(world);
        TodoResponse todo = findByTitle(title);
        APIResponse response = world.api().delete(todosUrl() + "/" + todo.id());
        if (response.status() != 204) {
            throw new RuntimeException("Failed to delete todo: " + response.status() + " " + response.statusText());
        }
    }

    @Then("the API should return the todo {string} with status {string}")
    public void theApiShouldReturnTheTodoWithStatus(String title, String status) {
        ApiHelpers.ensureApi(world);
        TodoResponse found = findByTitle(title);
        if (!found.status().equals(status)) {
            throw new AssertionError("Expected status \"" + status + "\" but got \"" + found.status() + "\"");
        }
    }

    @Then("the API should return the todo {string} with priority {string}")
    public void theApiShouldReturnTheTodoWithPriority(String title, String priority) {
        ApiHelpers.ensureApi(world);
        TodoResponse found = findByTitle(title);
        if (!found.priority().equals(priority)) {
            throw new AssertionError("Expected priority \"" + priority + "\" but got \"" + found.priority() + "\"");
        }
    }

    @Then("the API should not return a todo {string}")
    public void theApiShouldNotReturnATodo(String title) {
        ApiHelpers.ensureApi(world);
        APIResponse response = world.api().get(todosUrl());
        requireOk(response, "list todos");
        List<TodoResponse> todos = world.api().jsonList(response, TodoResponse.class);
        boolean found = todos.stream().anyMatch(t -> t.title().equals(title));
        if (found) {
            throw new AssertionError("Todo \"" + title + "\" should not exist but was found");
        }
    }

    @Then("the API should return {int} todo(s)")
    public void theApiShouldReturnTodos(int count) {
        ApiHelpers.ensureApi(world);
        APIResponse response = world.api().get(todosUrl());
        requireOk(response, "list todos");
        List<TodoResponse> todos = world.api().jsonList(response, TodoResponse.class);
        if (todos.size() != count) {
            throw new AssertionError("Expected " + count + " todos but got " + todos.size());
        }
    }
}
